### ISSUE TITLE
Expanding command substitutions or other constructs interactively with tab

### ISSUE THREAD

**Issue by nhooyr** [fish-shell/fish-shell#5536]
See #5508

E.g. if I'm in my cmdline and I've got (my cursor is `|`):

```fish
$ echo (echo MEOW)|
```

Pressing tab should give me:

```fish
$ echo MEOW
```

This would be a really neat feature to have because it'd allow you to ensure your command substitutions or file globs expand to the correct files/paths before running the cmd.

**Comment by floam** [comment]
This is a cool idea. My first thought however is this would get really ugly with many command substitutions and globs. We're potentially talking about fish suddenly rendering the command line taller than the terminal window with multiple pages of resulting expansion.

**Comment by nhooyr** [comment]
We can open the users pager if it’s bigger than either the users window or 20 lines.

**Comment by mqudsi** [comment]
I have concerns about this, particularly because it may result in transforming a idempotent operation to a non-idempotent operation post-expansion, and it is hard to go back from expanded form to unexpanded form. `TAB` should be safe and non-destructive, current tab completions are strictly append-only and make no modifications to anything the user has already entered at the prompt.

That said, there is definite value in this. Perhaps some sort of interactive confirmation ("Press TAB again to expand") or something to prevent accidentally triggering this.

> We can open the users pager if it’s bigger than either the users window or 20 lines.

You lost me there. `$PAGER` is a read-only viewer, not an editor. Even if you meant `$EDITOR`, nowhere else in fish does normal interaction at the prompt result in such behavior, it's a huge paradigm shift.

**Comment by nhooyr** [comment]
> I have concerns about this, particularly because it may result in transforming a idempotent operation to a non-idempotent operation post-expansion, and it is hard to go back from expanded form to unexpanded form. TAB should be safe and non-destructive, current tab completions are strictly append-only and make no modifications to anything the user has already entered at the prompt.

I disagree. I think it's an intuitive extension of tab completion. Tab complete isn't just show me a list of options that I can select from with the given prefix, in pretty much every IDE/text editor, you see support for snippets and more powerful forms of completion that replace the completed text entirely.

This is a default behaviour in ZSH and while ZSH is certainly not the best shell to take inspiration from, especially from its defaults, it certainly aids the argument that it is intuitive.

> That said, there is definite value in this. Perhaps some sort of interactive confirmation ("Press TAB again to expand") or something to prevent accidentally triggering this.

I'd be against the second TAB, it'd make it annoying to use. I think making it as simple as possible to use is worth the tradeoff in it being accidentally triggered. Very few command substitutions I have ever written actually modify state, they usually just return some processed text.

At most, maybe a different key bind by default like alt+tab may would make sense.

> You lost me there. $PAGER is a read-only viewer, not an editor. Even if you meant $EDITOR, nowhere else in fish does normal interaction at the prompt result in such behavior, it's a huge paradigm shift.

Yes, I meant editor. Pressing `ALT+E` for me opens the current typed command in my `$EDITOR` so its not that different.

However, I don't think we should do that. This sort of expansion should only occur if the output is small, say less than 5 lines. If the expansion is longer, fish could throw an error and then copy their current command onto the next prompt exactly.

**Comment by mqudsi** [comment]
Thanks for further explaining, but you've not done much to allay my concerns besides arguing that this is a feature ZSH has. Idempotency is important, otherwise history loses its value if the same command doesn't repeat the same (intended) results down the line, and arguably all you gain here is just a *sometimes* less lengthy command by replacing an expression with its expanded value - which could just as easily result in output _longer_ than the expression in question itself. The fishy way of doing that would be to assign the output of an operation to a variable and then use that in subsequent commands.

> However, I don't think we should do that. This sort of expansion should only occur if the output is small, say less than 5 lines. If the expansion is longer, fish could throw an error and then copy their current command onto the next prompt exactly.

That is another reason not to incorporate such a feature. Behavior should be mentally consistent and should involve as few "exceptions" as possible. It has been proven to be mentally less taxing to remember something longer than to remember something that changes or takes alternating paths [citation needed].

> you see support for snippets and more powerful forms of completion that replace the completed text entirely.

In some editors/environments, yes. In fish, no (with the exception of abbreviations, which are very deterministic).

**Comment by nhooyr** [comment]
>  arguably all you gain here is just a sometimes less lengthy command by replacing an expression with its expanded value

Thats not the purpose of this feature. Its to be able to check interactively that the expansion produces what you intend it to.

> The fishy way of doing that would be to assign the output of an operation to a variable and then use that in subsequent commands.

Right now thats the fish way but thats not a good UX imo. Especially as you write a command interactively and then think to yourself, I should double check the result of this expansion. You need to copy and paste instead of just pressing tab.

> That is another reason not to incorporate such a feature. Behavior should be mentally consistent and should involve as few "exceptions" as possible. It has been proven to be mentally less taxing to remember something longer than to remember something that changes or takes alternating paths [citation needed].

If fish throws an error, there is no effective mental tax. You can just rely on fish to tell you when its too long and you should probably be writing the expansion to a file to view it.

> Idempotency is important, otherwise history loses its value if the same command doesn't repeat the same (intended) results down the line.

Not sure how this feature affects that. History doesn't have idempotency right now.

> In some editors/environments, yes. In fish, no (with the exception of abbreviations, which are very deterministic).

The fact that fish does not currently have more powerful completion doesn't mean it shouldn't. Given pretty much every modern editor does, I think thats a very strong argument for it.

**Comment by zanchey** [comment]
I thought we'd discussed this somewhere, but I can only find #751, which suggests the (safer) expand binding.

**Comment by nhooyr** [comment]
@zanchey what's the difference between this and #751? Seems the same to me.
